zebra-chain/src/commitments/sapling.rs

```diff
@@ -0,0 +1,197 @@
+//! Sapling note and value commitments and types.
+
+#[cfg(test)]
+mod arbitrary;
+#[cfg(test)]
+mod test_vectors;
+
+pub mod pedersen_hashes;
+
+use std::fmt;
+
+use bitvec::prelude::*;
+use rand_core::{CryptoRng, RngCore};
+
+use crate::{
+    keys::sapling::{find_group_hash, Diversifier, TransmissionKey},
+    serde_helpers,
+    types::amount::{Amount, NonNegative},
+};
+
+use pedersen_hashes::*;
+
+/// The randomness used in the Pedersen Hash for note commitment.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct CommitmentRandomness(jubjub::Fr);
+
+/// Note commitments for the output notes.
+#[derive(Clone, Copy, Deserialize, PartialEq, Serialize)]
+pub struct NoteCommitment(#[serde(with = "serde_helpers::AffinePoint")] pub jubjub::AffinePoint);
+
+impl fmt::Debug for NoteCommitment {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("NoteCommitment")
+            .field("u", &hex::encode(self.0.get_u().to_bytes()))
+            .field("v", &hex::encode(self.0.get_v().to_bytes()))
+            .finish()
+    }
+}
+
+impl From<[u8; 32]> for NoteCommitment {
+    fn from(bytes: [u8; 32]) -> Self {
+        Self(jubjub::AffinePoint::from_bytes(bytes).unwrap())
+    }
+}
+
+impl From<jubjub::ExtendedPoint> for NoteCommitment {
+    fn from(extended_point: jubjub::ExtendedPoint) -> Self {
+        Self(jubjub::AffinePoint::from(extended_point))
+    }
+}
+
+impl From<NoteCommitment> for [u8; 32] {
+    fn from(cm: NoteCommitment) -> [u8; 32] {
+        cm.0.to_bytes()
+    }
+}
+
+impl Eq for NoteCommitment {}
+
+impl NoteCommitment {
+    /// Generate a new _NoteCommitment_ and the randomness used to create it.
+    ///
+    /// We return the randomness because it is needed to construct a _Note_,
+    /// before it is encrypted as part of an _Output Description_.
+    ///
+    /// NoteCommit^Sapling_rcm (g*_d , pk*_d , v) :=
+    ///   WindowedPedersenCommit_rcm([1; 6] || I2LEBSP_64(v) || g*_d || pk*_d)
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#concretewindowedcommit
+    #[allow(non_snake_case)]
+    pub fn new<T>(
+        csprng: &mut T,
+        diversifier: Diversifier,
+        transmission_key: TransmissionKey,
+        value: Amount<NonNegative>,
+    ) -> (CommitmentRandomness, Self)
+    where
+        T: RngCore + CryptoRng,
+    {
+        // s as in the argument name for WindowedPedersenCommit_r(s)
+        let mut s: BitVec<Lsb0, u8> = BitVec::new();
+
+        // Prefix
+        s.append(&mut bitvec![1; 6]);
+
+        // Jubjub repr_J canonical byte encoding
+        // https://zips.z.cash/protocol/protocol.pdf#jubjub
+        //
+        // The `From<Diversifier>` impls for the `jubjub::*Point`s handles
+        // calling `DiversifyHash` implicitly.
+        let g_d_bytes = jubjub::AffinePoint::from(diversifier).to_bytes();
```

Comment:
Again you need to handle the case of an invalid diversifier. Either `NoteCommitment::new` has to return an Option, or Diversifier has to be restricted to diversifiers for which GroupHash returns non-⊥. (In the latter case you probably want to have Diversifier cache g_d.)

Reply:
This relates to @hdevalence 's latter question up on https://github.com/ZcashFoundation/zebra/pull/596#discussion_r466033908


A ⊥ value for a `Diversifier` is technically a valid value, so I think that should remain.
When consuming said values, the resulting value has to be a `Result` or `Option`, which is annoying for a constructor like `new()` here but you know, c'est la vie.

Reply:
@hdevalence @daira thoughts on [this](https://github.com/ZcashFoundation/zebra/pull/596/commits/b2dd674c6f6ff9a54d2ebb8a9d3ddb8804bd500e)?

Reply:
I answered on that commit.